src/cunumeric/stat/histogram_impl.h

```diff
@@ -0,0 +1,140 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+namespace cunumeric {
+namespace detail {
+
+template <typename exe_policy_t, typename array_t, typename bin_t>
+struct lower_bound_op_t {
+  lower_bound_op_t(exe_policy_t, bin_t const* p_bins, size_t n_intervs)
+    : p_bins_(p_bins), n_intervs_(n_intervs)  // CTAD
+  {
+  }
+  __host__ __device__ bool operator()(array_t left, bin_t right) const
+  {
+    // sentinel logic accounts for comparison
+    // against last bin's upper bound, when
+    // (<) is to be replaced by (<=):
+    //
+    auto sentinel = p_bins_[n_intervs_];
+    if constexpr (std::is_same_v<array_t, __half> && (!std::is_integral_v<bin_t>)) {
+      // upcast operands:
+      //
+      bin_t left_up  = static_cast<bin_t>(left);
+      bin_t right_up = static_cast<bin_t>(right);
+      if (left_up == sentinel && right_up == sentinel)
+        return true;
+      else
+        return left_up < right_up;
+    } else if (std::is_same_v<array_t, __half> && std::is_integral_v<bin_t>) {
+      // upcast sentinel:
+      //
+      array_t sentinel_up = static_cast<array_t>(sentinel);
+      if (left == sentinel_up && right == sentinel_up)
+        return true;
+      else
+        return left < right;
+    } else {
+      if (left == sentinel && right == sentinel)
+        return true;
+      else
+        return left < right;
+    }
+  }
+
+ private:
+  bin_t const* p_bins_;
+  size_t n_intervs_;
+};
+
+template <typename exe_policy_t, typename bin_t, typename weight_t>
+struct transform_op_t {
+  transform_op_t(exe_policy_t exe_pol, bin_t const* p_bins, weight_t* p_hist, size_t n_intervs)
+    : ptr_bins_(p_bins)
+  {
+    Sw_ = thrust::reduce(exe_pol, p_hist, p_hist + n_intervs);
+  }
+
+  __host__ __device__ weight_t operator()(size_t index, weight_t h_i)
+  {
+    auto d_i = ptr_bins_[index + 1] - ptr_bins_[index];
+
+#ifdef _ZERO_WIDTH_RET_ZERO_
+    if (d_i == 0)
+      return static_cast<weight_t>(0);
+    else
+#endif
+      return h_i / (Sw_ * d_i);
+  }
+
+ private:
+  bin_t const* ptr_bins_;
+  weight_t* ptr_h_;
+  weight_t Sw_{0};
+};
+
+template <typename exe_policy_t,
+          typename array_t,
+          typename bin_t,
+          template <typename...> typename alloc_t = allocator_t,
+          typename weight_t                       = array_t,
+          typename offset_t                       = size_t>
+void histogram_weights(exe_policy_t exe_pol,
+                       array_t* ptr_src,  // source array, a
+                       size_t n_samples,
+                       bin_t const* ptr_over,          // bins array,
+                       size_t n_intervals,             // |bins| - 1
+                       weight_t* ptr_hist,             // result; pre-allocated, sz = n_intervals
+                       weight_t* ptr_w     = nullptr,  // weights array, w
+                       cudaStream_t stream = nullptr)
```

Comment:
`histogram.cc` includes a bunch of thrust headers via `histogram_cpu.h`. This is how `cudaStream_t` becomes visible. The only thing required for `g++` to compile this is: `-I/usr/local/cuda/targets/x86_64-linux/include`, which I assume is there, since the build goes through. The CPU uses the `host` version of Thrust. So, it must include thrust headers.